### PR TITLE
Refactor client packages

### PR DESCRIPTION
**Pull request by dadgar**
This PR refactors the client into several smaller scoped packages.

### CODE REVIEW

File: client/allocrunner/alloc_runner.go [hunk 1]
```diff
@@ -13,7 +12,10 @@ import (
 	"github.com/boltdb/bolt"
 	"github.com/hashicorp/go-multierror"
 	"github.com/hashicorp/nomad/client/allocdir"
+	"github.com/hashicorp/nomad/client/allocrunner/taskrunner"
 	"github.com/hashicorp/nomad/client/config"
+	consulApi "github.com/hashicorp/nomad/client/consul"
```
Comment (schmichael) [inline client/allocrunner/alloc_runner.go]: [brainstorming] If we were feeling brave and wanted to prepare for a service-discovery plugins future you could rename this package "discoapi" or similar. Other than the package name there's nothing Consul-specific about it: the methods and structs should be sufficient for any discovery backend.

File: client/allocrunner/alloc_runner.go [hunk 2]
```diff
@@ -70,7 +72,7 @@ type AllocRunner struct {
 	updateCh chan *structs.Allocation
 
 	vaultClient  vaultclient.VaultClient
-	consulClient ConsulServiceAPI
+	consulClient consulApi.ConsulServiceAPI
```
Comment (schmichael) [inline client/allocrunner/alloc_runner.go]: I think these vars would be named something like `discoClient discoAPI.Client` given my brainstorming above.

File: client/client.go [hunk 3]
```diff
@@ -507,10 +510,16 @@ func (c *Client) Shutdown() error {
 
 	// Destroy all the running allocations.
 	if c.config.DevMode {
+		var wg sync.WaitGroup
 		for _, ar := range c.getAllocRunners() {
-			ar.Destroy()
-			<-ar.WaitCh()
+			wg.Add(1)
+			go func(ar *allocrunner.AllocRunner) {
```
Comment (schmichael) [inline client/client.go]: Nice! Alternatively you could have done two loops:

```go
for _, ar := range c.getAllocRunners() {
  ar.Destroy()
}

for _, ar := range c.getAllocRunners() {
  <-ar.WaitCh()
}
```

File: client/gc_test.go [hunk 4]
```diff
@@ -451,19 +485,23 @@ func TestAllocGarbageCollector_UsageBelowThreshold(t *testing.T) {
 
 func TestAllocGarbageCollector_UsedPercentThreshold(t *testing.T) {
 	t.Parallel()
-	logger := testLogger()
```
Comment (schmichael) [inline client/gc_test.go]: Did you remove the testLogger func too? I may have missed that.

Comment (dadgar) [reply client/gc_test.go]: Went through the entire code base and removed all calls to log.New in tests

**Pull request merged**